include/nbl/builtin/hlsl/workgroup/bitonic_sort.hlsl

```diff
@@ -0,0 +1,253 @@
+        comparator_t comp;
+
+        // PHASE 1: Sub-sorts in chunks of size WorkgroupSize*2
+        accessor_adaptors::Offset<Accessor> offsetAccessor;
+        offsetAccessor.accessor = accessor;
+
+        const uint32_t numSub = TotalElements / ElementsPerSimpleSort;
+
+        [unroll]
+        for (uint32_t sub = 0; sub < numSub; sub++)
+        {
+            if (sub)
+                sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+
+                offsetAccessor.offset = sub * ElementsPerSimpleSort;
+
+                // Call E=1 workgroup sort
+                BitonicSort<simple_config_t, device_capabilities>::template __call(offsetAccessor, sharedmemAccessor);
+        }
+        sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+
+        // PHASE 2: Reverse odd-indexed chunks to form bitonic sequences
+        const uint32_t simpleLog = hlsl::findMSB(ElementsPerSimpleSort - 1) + 1u;
+        [unroll]
+        for (uint32_t sub = 1; sub < numSub; sub += 2)
+        {
+            offsetAccessor.offset = sub * ElementsPerSimpleSort;
+            [unroll]
+            for (uint32_t strideLog = simpleLog - 1u; strideLog + 1u > 0u; strideLog--)
+            {
+                const uint32_t stride = 1u << strideLog;
+                [unroll]
+                for (uint32_t virtualThreadID = threadID; virtualThreadID < ElementsPerSimpleSort / 2; virtualThreadID += WorkgroupSize)
+                {
+                    const uint32_t loIx = (((virtualThreadID & (~(stride - 1u))) << 1u) | (virtualThreadID & (stride - 1u))) + offsetAccessor.offset;
+                    const uint32_t hiIx = loIx | stride;
+
+                    nbl::hlsl::pair<key_t, value_t> lopair, hipair;
+                    accessor.template get<nbl::hlsl::pair<key_t, value_t> >(loIx, lopair);
+                    accessor.template get<nbl::hlsl::pair<key_t, value_t> >(hiIx, hipair);
+
+                    swap(lopair, hipair);
+
+                    accessor.template set<nbl::hlsl::pair<key_t, value_t> >(loIx, lopair);
+                    accessor.template set<nbl::hlsl::pair<key_t, value_t> >(hiIx, hipair);
+                }
+                sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+                }
+            }
+
+            // PHASE 3: Global memory bitonic merge
+            const uint32_t totalLog = hlsl::findMSB(TotalElements - 1) + 1u;
+            [unroll]
+            for (uint32_t blockLog = simpleLog + 1u; blockLog <= totalLog; blockLog++)
+            {
+                const uint32_t k = 1u << blockLog;
+                [unroll]
+                for (uint32_t strideLog = blockLog - 1u; strideLog + 1u > 0u; strideLog--)
+                {
+                    const uint32_t stride = 1u << strideLog;
+                    [unroll]
+                    for (uint32_t virtualThreadID = threadID; virtualThreadID < TotalElements / 2; virtualThreadID += WorkgroupSize)
```

Comment:
this won't unroll, but this will
```cpp
for (uint32_t baseVirtualThread=0; baseVirtualThread<HalfTotalElements; baseVirtualThread+=WorkgroupSize)
{
  const uint32_t virtualThreadID = baseVirtualThread+threadID;
```

the compiler now see that the loop condition is invariant (otherwise you have a dep on `threadID` and netiher DXC or SPIR-V opt can do the long and ardous analysis to see if your workgroup count is a multiple and the loop condition is actually uniform for whole group always)